### PR TITLE
binlog: support `binlog.pump.config` configurations for pump and drainer 

### PR DESCRIPTION
**Pull request by weekface**
### What problem does this PR solve? 

- support `binlog.pump.config` and `binlog.drainer.config` configurations for pump and drainer 
- support pump and drainer configuration rollout 

### Check List 

Tests 

 - E2E test
 - Stability test

Code changes

 - Has Helm charts change
 - Has Go code change
 - Has documents change

Side effects

 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:
 
 ```release-note
support `binlog.pump.config` and `binlog.drainer.config` configurations for pump and drainer 
 ```

### CODE REVIEW

File: charts/tidb-cluster/templates/pump-statefulset.yaml [hunk 1]
```diff
@@ -50,9 +50,13 @@ spec:
       volumes:
       - name: config
         configMap:
+        {{- if .Values.enableConfigMapRollout }}
+          name: {{ template "cluster.name" . }}-pump-{{ template "pump-configmap.data-digest" . }}
```
Comment (weekface) [inline charts/tidb-cluster/templates/pump-statefulset.yaml]: pump configration rollout @aylei PTAL

File: charts/tidb-cluster/templates/drainer-statefulset.yaml [hunk 2]
```diff
@@ -50,9 +50,13 @@ spec:
       volumes:
       - name: config
         configMap:
+        {{- if .Values.enableConfigMapRollout }}
+          name: {{ template "cluster.name" . }}-drainer-{{ template "drainer-configmap.data-digest" . }}
```
Comment (weekface) [inline charts/tidb-cluster/templates/drainer-statefulset.yaml]: drainer configration rollout @aylei PTAL

File: charts/tidb-cluster/values.yaml [hunk 3]
```diff
@@ -398,13 +398,17 @@ binlog:
...
+    # Please refer to https://github.com/pingcap/tidb-binlog/blob/master/cmd/pump/pump.toml for the default
+    # pump configurations (change to the tags of your pump version),
+    # just follow the format in the file and configure in the 'config' section
+    # as below if you want to customize any configuration.
+    config: |
+      gc = 7
+      heartbeat-interval = 2
+      [storage]
+      sync-log = true
+      stop-write-at-available-space = "100 gib"
```
Comment (aylei) [inline charts/tidb-cluster/values.yaml]:
```
# stop write when disk available space less then the configured size
# 42 MB -> 42000000, 42 mib -> 44040192
# default: 10 gib
# stop-write-at-available-space = "10 gib"
```
I think it should be 1 Gigabytes or less

File: charts/tidb-cluster/values.yaml [hunk 4]
```diff
@@ -417,39 +421,27 @@ binlog:
...
-    kafka: {}
-      # only need config one of zookeeper-addrs and kafka-addrs, will get kafka address if zookeeper-addrs is configed.
-      # zookeeperAddrs: "127.0.0.1:2181"
-      # kafkaAddrs: "127.0.0.1:9092"
-      # kafkaVersion: "0.8.2.0"
+    # Please refer to https://github.com/pingcap/tidb-binlog/blob/master/cmd/drainer/drainer.toml for the default
+    # drainer configurations (change to the tags of your drainer version),
+    # just follow the format in the file and configure in the 'config' section
+    # as below if you want to customize any configuration.
+    config: |
```
Comment (aylei) [inline charts/tidb-cluster/values.yaml]: This causes the old configuration key invalid, which is a in-compatible change

**Comment by aylei** [comment]
I think rollout the binlog configuration is not a small change, can we put this to 1.1

File: charts/tidb-cluster/values.yaml (hunk 4 above)
Comment (weekface) [reply charts/tidb-cluster/values.yaml]: yes, this is an in-compatible change like: https://github.com/pingcap/tidb-operator/pull/638, i think we should do it before GA. @tennix

File: charts/tidb-cluster/values.yaml (hunk 3 above)
Comment (weekface) [reply charts/tidb-cluster/values.yaml]: Yes, this is available space size, using default.

File: charts/tidb-cluster/templates/scripts/_start_drainer.sh.tpl [hunk 5]
```diff
@@ -26,6 +26,7 @@ done
 
 /drainer \
 -L={{ .Values.binlog.drainer.logLevel | default "info" }} \
+-pd-urls={{ template "cluster.name" . }}-pd:2379 \
```
Comment (xiaojingchen) [inline charts/tidb-cluster/templates/scripts/_start_drainer.sh.tpl]: add `http://` scheme

File: charts/tidb-cluster/templates/scripts/_start_pump.sh.tpl [hunk 6]
```diff
@@ -1,5 +1,6 @@
 set -euo pipefail
 /pump \
+-pd-urls={{ template "cluster.name" . }}-pd:2379 \
```
Comment (xiaojingchen) [inline charts/tidb-cluster/templates/scripts/_start_pump.sh.tpl]: ditto

File: charts/tidb-cluster/values.yaml (hunk 4 above)
Comment (DanielZhangQD) [reply charts/tidb-cluster/values.yaml]: What about we keep these two ways, but make it clear in the document that we do not maintain the config for the original way and users should migrate to the new way if they want to configure any new parameters that are not exposed in values.yaml.

**Comment by weekface** [comment]
@aylei @xiaojingchen @tennix PTAL again

File: charts/tidb-cluster/values.yaml (hunk 4 above)
Comment (DanielZhangQD) [reply charts/tidb-cluster/values.yaml]: Have we decided to make this change incompatible?

Comment (weekface) [reply charts/tidb-cluster/values.yaml]: What do you think? @tennix

Comment (weekface) [reply charts/tidb-cluster/values.yaml]: Agree with @DanielZhangQD we should not make this incompatible change.

Comment (tennix) [reply charts/tidb-cluster/values.yaml]: I think we can keep both for binlog configurations to keep compatible.

**Comment by weekface** [comment]
@aylei @DanielZhangQD @tennix 

This PR changed to support both configurations. PTAL again.

File: charts/tidb-cluster/values.yaml (hunk 4 above)
Comment (weekface) [reply charts/tidb-cluster/values.yaml]: comments addressed.

File: charts/tidb-cluster/values.yaml [hunk 7]
```diff
@@ -423,6 +423,17 @@ binlog:
...
+    # Please refer to https://github.com/pingcap/tidb-binlog/blob/master/cmd/pump/pump.toml for the default
+    # pump configurations (change to the tags of your pump version),
+    # just follow the format in the file and configure in the 'config' section
+    # as below if you want to customize any configuration.
+    # config: |
+    #   gc = 7
+    #   heartbeat-interval = 2
+    #   [storage]
+    #   sync-log = true
+    #   stop-write-at-available-space = "10 gib"
```
Comment (tennix) [inline charts/tidb-cluster/values.yaml]:
```suggestion
    #   stop-write-at-available-space = "10Gi"
```

**Pull request merged**